internal/venice-client-common/src/main/java/com/linkedin/venice/compression/VeniceCompressor.java

```diff
@@ -1,43 +1,102 @@
 package com.linkedin.venice.compression;
 
+import com.linkedin.venice.exceptions.VeniceException;
 import com.linkedin.venice.utils.ByteUtils;
 import java.io.Closeable;
 import java.io.IOException;
 import java.io.InputStream;
 import java.nio.ByteBuffer;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
 
 
 public abstract class VeniceCompressor implements Closeable {
   protected static final int SCHEMA_HEADER_LENGTH = ByteUtils.SIZE_OF_INT;
   private final CompressionStrategy compressionStrategy;
+  private boolean isClosed = false;
+  private final ReentrantReadWriteLock readWriteLock = new ReentrantReadWriteLock();
 
   protected VeniceCompressor(CompressionStrategy compressionStrategy) {
     this.compressionStrategy = compressionStrategy;
   }
 
-  public abstract byte[] compress(byte[] data) throws IOException;
+  interface CompressionRunnable<R> {
+    R run() throws IOException;
+  }
+
+  private <R> R executeWithSafeGuard(CompressionRunnable<R> runnable) throws IOException {
+    readWriteLock.readLock().lock();
+    try {
+      if (isClosed) {
+        throw new VeniceException(getCompressionStrategy() + " has been closed");
+      }
+      return runnable.run();
+    } catch (IOException e) {
```

Comment:
can we capture all exception here ? It can throw `java.lang.IllegalStateException`

Reply:
It doesn't matter.
The key point here is to make sure  the race condition between `close` and `compress/decompress` operations doesn't happen.

Reply:
And `IllegalStateException` is a `RuntimeException`.